### ISSUE TITLE
Processes with inputs

### ISSUE THREAD

**Issue by hunse** [nengo/nengo#935]
Right now, the interface for running processes outside of a model (i.e., the `Process.run` and `Process.run_steps` functions) assumes that the process does not take any input. We'd like to have Processes with inputs, for example the Synapse classes should be derived from Process. What should the interface look like?

Some questions are:
- Do we want to support processes that may or may not have inputs, or will everything always be one type or the other? I currently can't think of any type of Process that I'd want to be both.
- If all Processes will either have inputs or not, should we have two subclasses? Like `ProcessWithInput` and `ProcessWithoutInput` (certainly we can think of better names, but be careful, because something like `InputProcess` is ambiguous, since Processes without inputs are typically used as inputs to a model).
- Could we come up with one interface that will serve both situations? One option would be to have all separate functions that would all take care of specific situations (`run_steps` running inputless for n steps, `run_time` running inputless for t seconds, `run_input` running with an input where the length of the input determines the runtime). We could also try to have one `run` function that tries to determine what to do based on the arguments it gets (e.g. given an array, run with input, given a float, run for t seconds without input) but that kind of programming model can get confusing and have insidious bugs.

**Comment by tbekolay** [comment]
Thanks for starting the discussion! I was thinking about this a while back, but didn't think about it enough to make an issue for it.

I agree that I can't think of a process that I would want to be both; at the same time, I am not a big fan of having two subclasses as it's hard to think of what they would be called, which indicates to me that the subclassing isn't appropriate.

Having separate functions, then, seems like the best solution to me. How about using the term "sample" for inputless processes? I don't know the right place to look to for terminology, but for distributions (which are similar in many ways to processes) we use sample. So, processes could have `run` and `run_steps` for running a processes with input, and `sample` and `sample_steps` for running a process without input.

Another term we might consider is "apply" ... I'm not sure where that would fit in, but for filters it makes sense to apply the filter to the input. So it could be apply / sample, though I'm just throwing those out there, not very invested in them.

**Comment by hunse** [comment]
We don't need `run` and `run_steps` for with input, since the input defines the runtime.

**Comment by hunse** [comment]
One other question is how to set the `size_out` for Processes with inputs. Some might have a fixed `size_out` no matter the `size_in`, for example a Process that computes the inner product of a vector with itself, and others might have a `size_out` dependent on the `size_in`, for example synapses.

**Comment by tbekolay** [comment]
I agree that sample is often used with random things (process is also normally used with random things) but it's also used with any sensor data... recording devices in neuroscience experiments have a sampling rate that defines their timestep. Not that we have to use sample, just giving a counterexample :)

> One other question is how to set the size_out for Processes with inputs

Good question... feels like it should be either a value or a function of size_in, which means it should be a function of size_in and processes that don't care about size_in can ignore it. Which is one other way in which processes that take in inputs differ from those that don't. So maybe that sways me a bit to the multiple subclasses side.

Perhaps it would help to think of names for the process with input and process without input. Process feels good for things without input; for those with input... maybe `Filter`? It does feel right that Synapses would be a subclass of Filter, but `filter` is a builtin which makes it a bit less attractive. Can't think of any other options right now...

This might be worth discussing / brainstorming at the dev meeting tomorrow! If there isn't much GUI stuff happening.

**Comment by hunse** [comment]
`Filter` to me is a particular type of process with input that has the same size output as input. Though I guess that's not totally true; I would call convolution (at least how it's used in neural networks) a type of filtering, and it can have different size outputs from inputs. But I still think `Filter` is a bit too specific. What about something like multiplicative noise, that takes an input and scales it by a random variable? I wouldn't call that a filter. A more general term would be System, but that's *very* general.

**Comment by tbekolay** [comment]
Next actions:

1. Go back to process discussion to other ways of making processes.
2. Check out @hunse's other branch that has `run_input`.
3. Brainstorm other names for things (`run_with_input`, `apply`, `process`, ???) (`Process` / `Filter` / `System`).

**Comment by hunse** [comment]
I was just looking at making Synapses into Processes. One issue is right now, the Synapse `make_step` function takes `output` as an argument. The main benefit of this is we can operate on the output in place, saving making a temporary memory buffer. For first-order synapses (like Lowpass), this also means that we can save keeping an extra memory buffer around permanently, since all the state it needs is stored in `output`.

If we make Synapses into a subclass of Process, we'll have to streamline this. The most obvious option is just get rid of it, and just accept the extra memory usage. The code would be cleaner, and the extra memory not too impactful, and of course other backends are free to do what they want. This fits with our philosophy of the reference backend focusing on simplicity over performance. I would go this way.

The other option is to allow all processes to take an optional output buffer that their step function can write to. The main downside to this is it would be up to users to implement this in their `make_step` function if they're making a new process. They could choose to not support this functionality, as it would be unnecessary for most Processes, but then we have to be careful about where we use it.

I guess the third option is not to streamline it, and have Synapses be special. But that raises the question of why have them be Processes at all if we're just going to do that. I guess this option rests on whether there would still be significant Process code that Synapses could use, and I'm not sure that would be the case if they have a different `make_step` signature. Though we could keep `output` as a keyword argument, so they can still use `run_with_input` just making the extra buffer, but when we use them in the builder we pass the `output` argument to save on memory. Now that I say this, it doesn't sound as terrible. But it still does make Synapses special, and means that we have to keep specific builder code for building Synapses versus other Processes, which seems a bit strange.
